### ISSUE TITLE
Look-alike character `：` causes OneDrive remote files to be deleted on sync

### ISSUE THREAD

**Issue by ethack** [rclone/rclone#3128]
#### What is the problem you are having with rclone?
On the OneDrive (personal) backend files or directories with the `：` character cause issues with sync.  This is a unicode look-alike character for the standard colon `:`.  I use a program that auto-replaces a standard colon with the unicode `：` in some of my file and directory names and these files are causing problems.

I searched and found these issues that seem related, and I may just be hitting the same thing they are.
- https://github.com/ncw/rclone/issues/2920
- https://github.com/ncw/rclone/issues/2913
- https://github.com/ncw/rclone/issues/2633

Is my best option to just rename all my files to either use a regular colon `:` or not use either character at all?

#### What is your rclone version (output from `rclone version`)
Reproduced using both latest stable and beta versions.

```
rclone v1.47.0
- os/arch: linux/amd64
- go version: go1.12.4

rclone v1.47.0-015-gaba15f11-beta
- os/arch: linux/amd64
- go version: go1.12.4
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Debian 9.8 64-bit

####  Which cloud storage system are you using? (eg Google Drive)
OneDrive Personal

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
`rclone -vv sync test/ onedrive:test/`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

I created a test directory with the following files:
```
ethan@nas:~$ rclone tree test
/
├── test: two.txt
├── test：dir
│   └── four.txt
├── test：one.txt
└── three.txt
```

Notice that both `test/test：dir` and `test/test：one.txt` contain the character in question.  `test/test: two.txt` contains a normal colon. The other files are just there to show what happens to normal files.

I run `sync` the first time and all files are uploaded as expected.
```
ethack@nas:~$ rclone -vv sync test/ onedrive:test/
2019/04/23 17:31:29 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
2019/04/23 17:31:29 DEBUG : Using config file from "/home/ethack/.config/rclone/rclone.conf"
2019/04/23 17:31:30 INFO  : One drive root 'test': Waiting for checks to finish
2019/04/23 17:31:30 INFO  : One drive root 'test': Waiting for transfers to finish
2019/04/23 17:31:31 DEBUG : test: two.txt: Starting multipart upload
2019/04/23 17:31:31 DEBUG : test：one.txt: Starting multipart upload
2019/04/23 17:31:31 DEBUG : three.txt: Starting multipart upload
2019/04/23 17:31:32 DEBUG : test：dir/four.txt: Starting multipart upload
2019/04/23 17:31:32 DEBUG : test：dir/four.txt: Uploading segment 0/17 size 17
2019/04/23 17:31:32 DEBUG : three.txt: Uploading segment 0/18 size 18
2019/04/23 17:31:32 DEBUG : test: two.txt: Uploading segment 0/16 size 16
2019/04/23 17:31:32 DEBUG : test：one.txt: Uploading segment 0/16 size 16
2019/04/23 17:31:33 INFO  : test：dir/four.txt: Copied (new)
2019/04/23 17:31:33 INFO  : test: two.txt: Copied (new)
2019/04/23 17:31:33 INFO  : test：one.txt: Copied (new)
2019/04/23 17:31:33 INFO  : three.txt: Copied (new)
2019/04/23 17:31:33 INFO  : Waiting for deletions to finish
2019/04/23 17:31:33 INFO  :
Transferred:            67 / 67 Bytes, 100%, 17 Bytes/s, ETA 0s
Errors:                 0
Checks:                 0 / 0, -
Transferred:            4 / 4, 100%
Elapsed time:        3.7s

2019/04/23 17:31:33 DEBUG : 20 go routines active
2019/04/23 17:31:33 DEBUG : rclone: Version "v1.47.0" finishing with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
```
![image](https://user-images.githubusercontent.com/1696711/56620695-681f9480-65ef-11e9-90d0-950002d557b3.png)

The output of `ls` shows the files exist but it shows a colon instead of the original `：`.
```
ethack@nas:~$ rclone ls onedrive:test/
       16 test: two.txt
       16 test:one.txt
       18 three.txt
       17 test:dir/four.txt
```

At this point I downloaded these files from the web interface and confirmed that their checksums matched the originals, though it appears that OneDrive is doing the auto-replacement of the colon `:` in `test: two.txt` on its own because in both the screenshot and the downloaded file, it is not a normal `:` but the `：` replacement character.

If I run `sync` again immediately after notice that the file and folder containing `：` are deleted. This is where the problem is.  I have a bunch of files that are continually getting deleted and then re-uploaded the next time `sync` is run.
```
ethack@nas:~$ rclone -vv sync test/ onedrive:test/
2019/04/23 17:34:09 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
2019/04/23 17:34:09 DEBUG : Using config file from "/home/ethack/.config/rclone/rclone.conf"
2019/04/23 17:34:10 DEBUG : test：one.txt: Starting multipart upload
2019/04/23 17:34:10 DEBUG : test: two.txt: Size and modification time the same (differ by -405.615991ms, within tolerance 1s)
2019/04/23 17:34:10 DEBUG : test: two.txt: Unchanged skipping
2019/04/23 17:34:10 DEBUG : three.txt: Size and modification time the same (differ by -964.796824ms, within tolerance 1s)
2019/04/23 17:34:10 DEBUG : three.txt: Unchanged skipping
2019/04/23 17:34:11 DEBUG : test：dir/four.txt: Starting multipart upload
2019/04/23 17:34:11 INFO  : One drive root 'test': Waiting for checks to finish
2019/04/23 17:34:11 INFO  : One drive root 'test': Waiting for transfers to finish
2019/04/23 17:34:11 DEBUG : test：dir/four.txt: Uploading segment 0/17 size 17
2019/04/23 17:34:11 DEBUG : test：one.txt: Uploading segment 0/16 size 16
2019/04/23 17:34:11 INFO  : test：dir/four.txt: Copied (new)
2019/04/23 17:34:12 INFO  : test：one.txt: Copied (new)
2019/04/23 17:34:12 INFO  : Waiting for deletions to finish
2019/04/23 17:34:12 INFO  : test:dir/four.txt: Deleted
2019/04/23 17:34:12 INFO  : test:one.txt: Deleted
2019/04/23 17:34:12 DEBUG : test:dir: Removing directory
2019/04/23 17:34:13 DEBUG : One drive root 'test': deleted 1 directories
2019/04/23 17:34:13 INFO  :
Transferred:            33 / 33 Bytes, 100%, 7 Bytes/s, ETA 0s
Errors:                 0
Checks:                 4 / 4, 100%
Transferred:            2 / 2, 100%
Elapsed time:        4.5s

2019/04/23 17:34:13 DEBUG : 15 go routines active
2019/04/23 17:34:13 DEBUG : rclone: Version "v1.47.0" finishing with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
```

The folder is deleted even though the file it contains does not have any special characters in it.
![image](https://user-images.githubusercontent.com/1696711/56620807-cf3d4900-65ef-11e9-8360-31f9389c5053.png)

```
ethack@nas:~$ rclone ls onedrive:test/
       16 test: two.txt
       18 three.txt
```

Every other `sync` alternates uploading and deleting these files. The third and fourth runs below show a repeat of the above behavior.  After the third run all files exist as expected.  After the fourth the `：` file and folder are deleted.
```
ethack@nas:~$ rclone -vv sync test/ onedrive:test/
2019/04/23 17:34:09 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
2019/04/23 17:34:09 DEBUG : Using config file from "/home/ethack/.config/rclone/rclone.conf"
2019/04/23 17:34:10 DEBUG : test：one.txt: Starting multipart upload
2019/04/23 17:34:10 DEBUG : test: two.txt: Size and modification time the same (differ by -405.615991ms, within tolerance 1s)
2019/04/23 17:34:10 DEBUG : test: two.txt: Unchanged skipping
2019/04/23 17:34:10 DEBUG : three.txt: Size and modification time the same (differ by -964.796824ms, within tolerance 1s)
2019/04/23 17:34:10 DEBUG : three.txt: Unchanged skipping
2019/04/23 17:34:11 DEBUG : test：dir/four.txt: Starting multipart upload
2019/04/23 17:34:11 INFO  : One drive root 'test': Waiting for checks to finish
2019/04/23 17:34:11 INFO  : One drive root 'test': Waiting for transfers to finish
2019/04/23 17:34:11 DEBUG : test：dir/four.txt: Uploading segment 0/17 size 17
2019/04/23 17:34:11 DEBUG : test：one.txt: Uploading segment 0/16 size 16
2019/04/23 17:34:11 INFO  : test：dir/four.txt: Copied (new)
2019/04/23 17:34:12 INFO  : test：one.txt: Copied (new)
2019/04/23 17:34:12 INFO  : Waiting for deletions to finish
2019/04/23 17:34:12 INFO  : test:dir/four.txt: Deleted
2019/04/23 17:34:12 INFO  : test:one.txt: Deleted
2019/04/23 17:34:12 DEBUG : test:dir: Removing directory
2019/04/23 17:34:13 DEBUG : One drive root 'test': deleted 1 directories
2019/04/23 17:34:13 INFO  :
Transferred:            33 / 33 Bytes, 100%, 7 Bytes/s, ETA 0s
Errors:                 0
Checks:                 4 / 4, 100%
Transferred:            2 / 2, 100%
Elapsed time:        4.5s

2019/04/23 17:34:13 DEBUG : 15 go routines active
2019/04/23 17:34:13 DEBUG : rclone: Version "v1.47.0" finishing with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]

ethack@nas:~$ rclone -vv sync test/ onedrive:test/
2019/04/23 17:35:39 DEBUG : rclone: Version "v1.47.0" starting with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
2019/04/23 17:35:39 DEBUG : Using config file from "/home/ethack/.config/rclone/rclone.conf"
2019/04/23 17:35:40 DEBUG : test：one.txt: Starting multipart upload
2019/04/23 17:35:40 DEBUG : test: two.txt: Size and modification time the same (differ by -405.615991ms, within tolerance 1s)
2019/04/23 17:35:40 DEBUG : test: two.txt: Unchanged skipping
2019/04/23 17:35:40 DEBUG : three.txt: Size and modification time the same (differ by -964.796824ms, within tolerance 1s)
2019/04/23 17:35:40 DEBUG : three.txt: Unchanged skipping
2019/04/23 17:35:40 INFO  : One drive root 'test': Waiting for checks to finish
2019/04/23 17:35:40 INFO  : One drive root 'test': Waiting for transfers to finish
2019/04/23 17:35:40 DEBUG : test：one.txt: Uploading segment 0/16 size 16
2019/04/23 17:35:40 INFO  : test：one.txt: Copied (new)
2019/04/23 17:35:40 DEBUG : test：dir/four.txt: Starting multipart upload
2019/04/23 17:35:41 DEBUG : test：dir/four.txt: Uploading segment 0/17 size 17
2019/04/23 17:35:42 INFO  : test：dir/four.txt: Copied (new)
2019/04/23 17:35:42 INFO  : Waiting for deletions to finish
2019/04/23 17:35:42 INFO  :
Transferred:            33 / 33 Bytes, 100%, 10 Bytes/s, ETA 0s
Errors:                 0
Checks:                 2 / 2, 100%
Transferred:            2 / 2, 100%
Elapsed time:        3.1s

2019/04/23 17:35:42 DEBUG : 8 go routines active
2019/04/23 17:35:42 DEBUG : rclone: Version "v1.47.0" finishing with parameters ["rclone" "-vv" "sync" "test/" "onedrive:test/"]
```

**Comment by ncw** [comment]
> At this point I downloaded these files from the web interface and confirmed that their checksums matched the originals, though it appears that OneDrive is doing the auto-replacement of the colon `:` in `test: two.txt` on its own because in both the screenshot and the downloaded file, it is not a normal `:` but the `：` replacement character.

Rclone does [translation of reserved characters](https://github.com/ncw/rclone/blob/master/backend/onedrive/replace.go) for onedrive, so if you are on a unix system you can have a local file called `test:colon` and it will be uploaded as `test:colon` but downloaded again as `test:colon`.

So I think what is happening is when you have a local file with the unicode replacement `:` in it, rclone uploads that as-is.  However that really represents a file with a normal `:` so when the sync is run again, rclone sees an excess file and deletes it.

@B4dM4n is part of the way through a plan to revamp how this all works (see #2955) so that the replacement characters rclone uses will be escaped if found in the local file system.

So if you can use a different character then this will fix the problem.

**Comment by ethack** [comment]
Thank you, I will rename my files to use something else.

**Comment by imfms** [comment]
full-width-character '？'(not ?) also have this problem.

**Comment by ncw** [comment]
I've merged the fix for this to master now which means it is in [the latest beta](https://beta.rclone.org) and will be released in v1.50
